diff --git a/lwip/arch/gc/netif/gcif.c b/lwip/arch/gc/netif/gcif.c
index b38c3f1..f65d09d 100644
--- a/lwip/arch/gc/netif/gcif.c
+++ b/lwip/arch/gc/netif/gcif.c
@@ -421,7 +421,7 @@ static void __bba_reset()
 
 static void __bba_recv_init()
 {
-	bba_out8(BBA_NCRB,(BBA_NCRB_AB|BBA_NCRB_CA|BBA_NCRB_2_PACKETS_PER_INT));
+	bba_out8(BBA_NCRB,(BBA_NCRB_AB|BBA_NCRB_CA|BBA_NCRB_1_PACKET_PER_INT));
 	bba_out8(BBA_SI_ACTRL2,0x74);
 	bba_out8(BBA_RXINTT, 0x00);
 	bba_out8(BBA_RXINTT+1, 0x06); /* 0x0600 = 61us */
