diff --git a/lwip/core/tcp_in.c b/lwip/core/tcp_in.c
index 8aca515..d50ba79 100644
--- a/lwip/core/tcp_in.c
+++ b/lwip/core/tcp_in.c
@@ -696,7 +696,7 @@ tcp_receive(struct tcp_pcb *pcb)
     } else
       /*if (TCP_SEQ_LT(pcb->lastack, ackno) &&
         TCP_SEQ_LEQ(ackno, pcb->snd_max)) { */
-      if (TCP_SEQ_BETWEEN(ackno, pcb->lastack+1, pcb->snd_max)){
+      if (TCP_SEQ_BETWEEN(ackno, pcb->lastack+1, pcb->snd_nxt)){
       /* We come here when the ACK acknowledges new data. */
       
       /* Reset the "IN Fast Retransmit" flag, since we are no longer
@@ -780,7 +780,7 @@ tcp_receive(struct tcp_pcb *pcb)
     while (pcb->unsent != NULL &&
            /*TCP_SEQ_LEQ(ntohl(pcb->unsent->tcphdr->seqno) + TCP_TCPLEN(pcb->unsent), ackno) &&
              TCP_SEQ_LEQ(ackno, pcb->snd_max)*/
-           TCP_SEQ_BETWEEN(ackno, ntohl(pcb->unsent->tcphdr->seqno) + TCP_TCPLEN(pcb->unsent), pcb->snd_max)
+           TCP_SEQ_BETWEEN(ackno, ntohl(pcb->unsent->tcphdr->seqno) + TCP_TCPLEN(pcb->unsent), pcb->snd_nxt)
            ) {
       LWIP_DEBUGF(TCP_INPUT_DEBUG, ("tcp_receive: removing %"U32_F":%"U32_F" from pcb->unsent\n",
                                     ntohl(pcb->unsent->tcphdr->seqno), ntohl(pcb->unsent->tcphdr->seqno) +
@@ -796,10 +796,6 @@ tcp_receive(struct tcp_pcb *pcb)
         LWIP_ASSERT("tcp_receive: valid queue length",
           pcb->unacked != NULL || pcb->unsent != NULL);
       }
-
-      if (pcb->unsent != NULL) {
-        pcb->snd_nxt = htonl(pcb->unsent->tcphdr->seqno);
-      }
     }
     /* End of ACK for new data processing. */
 
diff --git a/lwip/core/tcp_out.c b/lwip/core/tcp_out.c
index 754f649..20c9fc0 100644
--- a/lwip/core/tcp_out.c
+++ b/lwip/core/tcp_out.c
@@ -376,7 +376,7 @@ tcp_output(struct tcp_pcb *pcb)
   struct pbuf *p;
   struct tcp_hdr *tcphdr;
   struct tcp_seg *seg, *useg;
-  u32_t wnd;
+  u32_t wnd, snd_nxt;
 #if TCP_CWND_DEBUG
   s16_t i = 0;
 #endif /* TCP_CWND_DEBUG */
@@ -477,9 +477,9 @@ tcp_output(struct tcp_pcb *pcb)
     }
 
     tcp_output_segment(seg, pcb);
-    pcb->snd_nxt = ntohl(seg->tcphdr->seqno) + TCP_TCPLEN(seg);
-    if (TCP_SEQ_LT(pcb->snd_max, pcb->snd_nxt)) {
-      pcb->snd_max = pcb->snd_nxt;
+    snd_nxt = ntohl(seg->tcphdr->seqno) + TCP_TCPLEN(seg);
+    if (TCP_SEQ_LT(pcb->snd_nxt, snd_nxt)) {
+      pcb->snd_nxt = snd_nxt;
     }
     /* put segment on unacknowledged list if length > 0 */
     if (TCP_TCPLEN(seg) > 0) {
@@ -633,7 +633,6 @@ tcp_rexmit_rto(struct tcp_pcb *pcb)
   /* unacked queue is now empty */
   pcb->unacked = NULL;
 
-  pcb->snd_nxt = ntohl(pcb->unsent->tcphdr->seqno);
   /* increment number of retransmissions */
   ++pcb->nrtx;
 
@@ -659,8 +658,6 @@ tcp_rexmit(struct tcp_pcb *pcb)
   pcb->unsent = pcb->unacked;
   pcb->unacked = seg;
 
-  pcb->snd_nxt = ntohl(pcb->unsent->tcphdr->seqno);
-
   ++pcb->nrtx;
 
   /* Don't take any rtt measurements after retransmitting. */
